# Copyright 2026 Qore Technologies, s.r.o.; BSD-3-Clause.
# Private PDFium C ABI with distribution compiler/runtime and RPM flags.
--- a/BUILD.gn
+++ b/BUILD.gn
@@ -127,7 +127,7 @@
 }
 
 config("pdfium_implementation_config") {
-  defines = [ "FPDF_IMPLEMENTATION" ]
+  defines = [ "FPDF_IMPLEMENTATION", "FPDF_SHARED" ]
   visibility = [ ":pdfium_public_headers" ]
 }
 
@@ -226,8 +226,11 @@
   ]
 }
 
-component("pdfium") {
-  output_name = "pdfium"
+shared_library("pdfium") {
+  output_name = "pdfium-qore148"
+  output_extension = "so.0"
+  ldflags = [ "-Wl,--version-script=" + rebase_path("//pdfium-exports.map", root_build_dir) ]
+  inputs = [ "//pdfium-exports.map" ]
   libs = []
   configs += [ ":pdfium_strict_config" ]
   public_configs = [ ":pdfium_public_config" ]
--- a/public/fpdfview.h
+++ b/public/fpdfview.h
@@ -203,7 +203,7 @@
 // Dictionary value types.
 typedef int FPDF_OBJECT_TYPE;
 
-#if defined(COMPONENT_BUILD)
+#if defined(COMPONENT_BUILD) || defined(FPDF_SHARED)
 // FPDF_EXPORT should be consistent with |export| in the pdfium_fuzzer
 // template in testing/fuzzers/BUILD.gn.
 #if defined(WIN32)
--- a/build/config/compiler/BUILD.gn
+++ b/build/config/compiler/BUILD.gn
@@ -629,7 +629,7 @@
     # https://crbug.com/484082200 for background and https://crrev.com/c/7593035
     # for discussion.
     if (!is_wasm) {
-      cflags += [ "-fno-lifetime-dse" ]
+      # Supplied by rpm/buildflags.gni when supported by the distribution compiler.
     }
 
     # TODO(hans): Remove this once Clang generates better optimized debug info
@@ -674,7 +674,7 @@
     # It also causes segfault on Linux s390x:
     # https://github.com/llvm/llvm-project/issues/149511
     if (is_linux && use_lld && current_cpu != "arm" && current_cpu != "s390x") {
-      cflags += [ "-Wa,--crel,--allow-experimental-crel" ]
+      # Supplied by rpm/buildflags.gni when supported by the distribution compiler.
     }
   }
 
@@ -1049,6 +1049,11 @@
   if (current_cpu == "arm64") {
     defines += [ "__ARM_NEON__=1" ]
   }
+  import("//rpm/buildflags.gni")
+  cflags += qore_common_cflags + qore_cppflags + qore_extra_cflags
+  cflags_c += qore_cflags
+  cflags_cc += qore_cxxflags
+  ldflags += qore_ldflags
 }
 
 config("linker") {
@@ -1910,13 +1915,13 @@
       # sanitizer) is enabled, they then do expensive debug like operations. We
       # want to suppress this behaviour since we want to keep performance costs
       # as low as possible while having these checks.
-      "-fsanitize-ignore-for-ubsan-feature=array-bounds",
+      # Supplied by rpm/buildflags.gni when supported by the distribution compiler.
 
       # Because we've enabled array-bounds sanitizing we also want to suppress
       # the related warning about "unsafe-buffer-usage-in-static-sized-array",
       # since we know that the array bounds sanitizing will catch any out-of-
       # bounds accesses.
-      "-Wno-unsafe-buffer-usage-in-static-sized-array",
+      # Supplied by rpm/buildflags.gni when supported by the distribution compiler.
     ]
   }
 }
@@ -1934,7 +1939,7 @@
       # sanitizer) is enabled, they then do expensive debug like operations. We
       # want to suppress this behaviour since we want to keep performance costs
       # as low as possible while having these checks.
-      "-fsanitize-ignore-for-ubsan-feature=return",
+      # Supplied by rpm/buildflags.gni when supported by the distribution compiler.
     ]
   }
 }
--- a/build/config/clang/BUILD.gn
+++ b/build/config/clang/BUILD.gn
@@ -160,6 +160,11 @@
   if (!defined(invoker.libname) || is_wasm) {
     not_needed(invoker, "*")
     config(target_name) {
+    }
+  } else if (is_linux && clang_base_path != default_clang_base_path) {
+    config(target_name) {
+      import("//rpm/buildflags.gni")
+      libs = [ "$qore_compiler_rt_dir/libclang_rt.${invoker.libname}${qore_compiler_rt_suffix}.a" ]
     }
   } else {
     config(target_name) {
--- /dev/null
+++ b/pdfium-exports.map
@@ -0,0 +1,4 @@
+{
+  global: FPDF*; FORM_*; FSDK_*;
+  local: *;
+};

--- a/build/config/compiler/BUILD.gn
+++ b/build/config/compiler/BUILD.gn
@@ -1090,7 +1090,7 @@
       ldflags += [ "-Wl,--build-id=sha1" ]
     } else if (current_os != "aix" && current_os != "zos") {
       if (use_lld) {
-        ldflags += [ "-Wl,--build-id=fast" ]
+        ldflags += [ "-Wl,--build-id=sha1" ]
       } else {
         ldflags += [ "-Wl,--build-id" ]
       }
--- a/build/config/compiler/compiler.gni
+++ b/build/config/compiler/compiler.gni
@@ -325,7 +325,8 @@
 # If the platform uses stripped absolute paths by default, then we don't expose
 # it as a configuration option. If this is causing problems, please file a bug.
 if (strip_absolute_paths_from_debug_symbols_default) {
-  strip_absolute_paths_from_debug_symbols = true
+  # RPM debugedit rewrites absolute paths into reproducible debugsource paths.
+  strip_absolute_paths_from_debug_symbols = false
 } else {
   declare_args() {
     strip_absolute_paths_from_debug_symbols = false
--- a/build/config/compiler/BUILD.gn
+++ b/build/config/compiler/BUILD.gn
@@ -3258,8 +3258,9 @@
     # Disable debugger index for Android, which uses lldb. lldb can create (and
     # cache) its own index in about 20 seconds, so there's no need to pay the
     # price of an index on every link.
-    _add_symbol_index =
-        symbol_level == 2 && !is_apple && use_lld && !is_wasm && !is_android
+    # Distribution GDB reads complete DWARF 5 directly; its optional index
+    # formats are incompatible with this distribution Clang output.
+    _add_symbol_index = false
     if (_add_symbol_index) {
       if (use_dwarf5) {
         cflags += [ "-gpubnames" ]
